pallets/liquidity-pools-forwarder/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The router id does not have any forwarder info stored
+		ForwardInfoNotFound,
+		/// Failed to unwrap a message which should be a forwarded one
+		UnwrappingFailed,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(0)]
+		pub fn set_forwarder(
+			origin: OriginFor<T>,
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			RouterForwarding::<T>::insert(
+				&router_id,
+				ForwardInfo {
+					source_domain,
+					contract: forwarding_contract,
+				},
+			);
+
+			Self::deposit_event(Event::<T>::ForwarderSet {
+				router_id,
+				source_domain,
+				forwarding_contract,
+			});
+
+			Ok(())
+		}
+
+		/// Remove the forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(1)]
+		pub fn remove_forwarder(origin: OriginFor<T>, router_id: T::RouterId) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			RouterForwarding::<T>::take(&router_id)
+				.map(|info| {
+					Self::deposit_event(Event::<T>::ForwarderRemoved {
+						router_id,
+						source_domain: info.source_domain,
+						forwarding_contract: info.contract,
+					});
+				})
+				.ok_or(Error::<T>::ForwardInfoNotFound.into())
+		}
+	}
+
+	impl<T: Config> MessageSender for Pallet<T> {
+		type Message = T::Message;
+		type Middleware = T::RouterId;
+		type Origin = DomainAddress;
+
+		fn send(
+			router_id: T::RouterId,
+			origin: DomainAddress,
+			message: T::Message,
+		) -> DispatchResult {
+			let msg = RouterForwarding::<T>::get(&router_id)
+				.map(|info| {
+					T::Message::try_wrap_forward(info.source_domain, info.contract, message.clone())
+				})
+				.unwrap_or_else(|| {
+					ensure!(!message.is_forwarded(), Error::<T>::ForwardInfoNotFound);
+					Ok(message)
+				})?;
+
+			T::MessageSender::send(router_id, origin, msg)
+		}
+	}
+
+	impl<T: Config> MessageReceiver for Pallet<T> {
+		type Message = T::Message;
+		type Middleware = T::RouterId;
+		type Origin = DomainAddress;
+
+		fn receive(
+			router_id: T::RouterId,
+			domain_address: DomainAddress,
+			message: T::Message,
+		) -> DispatchResult {
+			// Message can be unwrapped iff it was forwarded
+			//
+			// NOTE: We can rely on EVM side to ensure forwarded messages are valid such
+			// that it suffices to filter for the existence of forwarding info
+			let lp_message = match (
+				RouterForwarding::<T>::get(&router_id).is_some(),
+				message.clone().unwrap_forwarded(),
+			) {
+				(true, Some((_domain, _contract, lp_message))) => Ok(lp_message),
```

Comment:
Because the origin of the LP message was not `_domain` (say Celo) but the `domain_address.domain()` (say Ethereum). The gateway checks in the allowlist, if it contains Ethereum's `domain_address.domain()` in the allowlist. I don't believe that Celo's `_domain` should be in the allowlist because there is no direct connection configured for Celo.

Reply:
Though now I remember why I initially went with the `RouterProvider`! We cannot guarantee that the `router_id` of the `receive` signature here is the same for which we have stored forwarding. Therefore, we actually need to iterate over all possible router ids!

cc @cdamian

Reply:
Mmm... Ok, here we have two points:

First, I think we need to read `_domain` or `_contract` somehow because if the other side forwarder implementation works like ours, it will never forward a message.

Second, if we need to use the forwarder domain, how do we receive the correct router?


Crazy idea. Could it makes sense that our chain never receives a Forwarder message? I mean, we're the final endpoint, right? The forwarder wrapper should have been unwrapped before reaching centrifuge. So, as a final endpoint, we should never receive a wrapped message.

Reply:
I think my above crazy idea is not as crazy, and it's related to [this thread](https://github.com/centrifuge/centrifuge-chain/pull/1971#discussion_r1720782323) above with Jeroen about not using the contract.

TLDR: I think we should not receive forwarded messages.

Reply:
> First, I think we need to read `_domain` or `_contract` somehow because if the other side forwarder implementation works like ours, it will never forward a message.

It won't work exactly the same, see [the spec](https://centrifuge.hackmd.io/@jeroen/Sy5MbJbzA?type=view#EVM-to-Centrifuge) and [this Slack response](https://kflabs.slack.com/archives/C06FRHCRQL9/p1723833895812759?thread_ts=1723801740.442379&cid=C06FRHCRQL9).

However, rereading the spec ECM to Centrifuge

> On Centrifuge
> 
> * Receives ForwardedMessage(source_chain=Gnosis, message=payload) from Ethereum
> * Executes payload, but with **source chain as Gnosis**

It appears to me we actually have to use the `source_domain` instead when handing over to the Gateway via `receive(...)` and require this `source_domain` to also be stored in the gateway allowlist.

Current flow Celo -> Ethereum -> Centrifuge is
* Forwarder checks whether message from Ethereum is a forwarded one 
* Assuming it is, extract the `source_domain` Celo and ensure forwarding storage is set up for Celo domain
* Hand over the extracted non-forward LP message to the gateway but with source domain and address pointing to the forwarder domain Ethereum

At least for handling tranche token transfers, this will be an issue, because it thinks the source domain was Ethereum. So I will fix this now.

Reply:
How will you then choose the correct `router_id` for the `source_domain`, @wischli ?

Reply:
See my above comment, the forwarder pallet needs to re-add the `RouterProvider` runtime type: https://github.com/centrifuge/centrifuge-chain/pull/1971#discussion_r1720905269

Reply:
But that way, we're breaking the proof checks, because all messages (even proofs) are redirected to the first found router.

cc @cdamian

Reply:
> But that way, we're breaking the proof checks, because all messages (even proofs) are redirected to the first found router, right?
> 
> cc @cdamian

It appears to me that this is a question of how to correclty configure/set up the storages. E.g. when calling `set_forwarding_info`, the correct router id needs to be stored which does not break our _first router_ invariant.

Moreover, this loop also breaks on the first found router.

Reply:
> `Forwarder X -> Ethereum -> Centrifuge` - message that comes through router A (pallet-axelar-router)
> `Forwarder X -> Ethereum -> Centrifuge` - proof that comes through router B (some other router pallet)

This scenario seems possible.

Reply:
OK, so in that case we can use the router ID that's provided to the forwarder in the receive call.

Reply:
Summary being:

>Forwarder X -> Ethereum -> Centrifuge - message that comes through router A (pallet-axelar-router)
this will call `receive` with router ID A

> Forwarder X -> Ethereum -> Centrifuge - proof that comes through router B (some other router pallet)
this will call `receive` with router ID B

Meaning that we shouldn't get the router ids for the domain provided when unwrapping the message?

Reply:
I see your point now. The current design wouldn't handle messages and their proofs correclty as everything is mapped to the same router id. So we are on the same page.

Revisiting our inbound setup and [my comment](https://github.com/centrifuge/centrifuge-chain/pull/1971#discussion_r1720939493) about correct forwarding configuration: `Router -> Serializer -> Forwarder -> Gateway`, it appears to me that we can actually rely on the correct `router_id` from the `receive` signature called by the `Router` and re-remove the `RouterProvider` from the forwarder pallet. 

E.g., when we store the forwarding info for a router id, we need to do that for every supported router. This way, the forwarder pallet has no impact on handling proof- vs non-proof messages and leaves this up to the Gateway based on the `router_id` which is passed all the way from the `MessageReceiver` impl of the corresponding router.

WDYT @cdamian @lemunozm

Reply:
> `Forwarder X -> Ethereum -> Centrifuge` - message that comes through router A (pallet-axelar-router)
> `Forwarder X -> Ethereum -> Centrifuge` - proof that comes through router B (some other router pallet)

This is the counterexample to the current implementation of the forwarder pallet.

Reply:
> I see your point now. The current design wouldn't handle messages and their proofs correclty as everything is mapped to the same router id. So we are on the same page.
> 
> Revisiting our inbound setup and [my comment](https://github.com/centrifuge/centrifuge-chain/pull/1971#discussion_r1720939493) about correct forwarding configuration: `Router -> Serializer -> Forwarder -> Gateway`, it appears to me that we can actually rely on the correct `router_id` from the `receive` signature called by the `Router` and re-remove the `RouterProvider` from the forwarder pallet.
> 
> E.g., when we store the forwarding info for a router id, we need to do that for every supported router. This way, the forwarder pallet has no impact on handling proof- vs non-proof messages and leaves this up to the Gateway based on the `router_id` which is passed all the way from the `MessageReceiver` impl of the corresponding router.
> 
> WDYT @cdamian @lemunozm

Completely agree.

Reply:
Copy pasting here the flows for a 2-router-setup you broke down to make this more explicit:

Flow of a non-forwarded message
```
1) Ethereum -> Centrifuge Router A -> LP Forwarder::receive(router A, domain address, message) -> noop -> LP Gateway::receive(router A, domain address, message)

2) Ethereum -> Centrifuge Router B -> LP Forwarder::receive(router A, domain address, message) - >noop -> LP Gateway::receive(router A, domain address, message)
```

Flow of a forwarded message
```
1) Celo(Ethereum) -> Centrifuge Router A -> LP Forwarder::receive(router A, domain address, message) -> unwrap message to (source_domain, contract, unwrapped_message) -> LP Gateway::receive(router A, (source domain, contract), unwrapped_message)

2) Celo(Ethereum) -> Centrifuge Router B -> LP Forwarder::receive(router B, domain address, message) -> unwrap message to (source_domain, contract, unwrapped_message) -> LP Gateway::receive(router B, (source domain, contract), unwrapped_message)
```

Reply:
Since we all have found consensus on this matter, here's the corresponding code: [e409fd5](https://github.com/centrifuge/centrifuge-chain/pull/1971/commits/e409fd59baf891910fd4a42a5ff7138cc1735e5f)

I really hope this was the last raid boss to slay and that we can merge this beast.

Reply:
I understand the second line of non-forwaded message is router B in all places right? @wischli